rfcs/20200612-stream-executor-c-api/C_API_versioning_strategy.md

```diff
@@ -1,226 +1,382 @@
+| **Authors** | Yi Situ (yisitu@google.com), Penporn Koanantakool (penporn@google.com), Anna Revinskaya (annarev@google.com) |
+| **Sponsor**   | Gunhan Gulsoy (gunan@google.com)                        |
+| **Updated**   | 2020-09-08                                              |
 
 In reply to a question on [PR #262](https://github.com/tensorflow/community/pull/262#issuecomment-653690654).
 
-# TensorFlow Versioning Strategy
-
-TensorFlow StreamExecutorInterface (SEI) uses struct_size for version checking. Struct members are not allowed to be removed or reordered.  Following are concrete examples of how TensorFlow remains compatible with plug-ins when functionality is added to or removed from StreamExecutorInterface. We will be using a simplified SE_Device as an example.
- 
-## When TensorFlow extends functionality
-### Backwards compatibility
-TensorFlow is compiled against a newer SEI header (v2), which has SE_Device extended with device_handle.
- 
-**Future TensorFlow compiled against StreamExecutorInterface v2**
-```cpp
-// StreamExecutorInterface header version 2
-typedef struct SE_Device {
-  size_t struct_size;
-  void* next;           // Always set to zero, reserved by TF for future use.
+StreamExecutor C API (SE C API) follows Semantic Versioning 2.0.0
+([semver](http://semver.org/)). Each release version has a format
+`MAJOR.MINOR.PATCH`, as outlined in [TensorFlow version compatibility](https://www.tensorflow.org/guide/versions#semantic_versioning_20).
+We also use `struct_size` to track compatibility.
+
+## Updating Guidelines
+This section outlines when to update version numbers specific to SE C API
+(`SE_MAJOR`, `SE_MINOR`, and `SE_PATCH`).
+
+### SE_MAJOR
+* Potentially backwards incompatible changes.
+* If a change is backwards incompatible, it requires an RFC because it will
+  break all current plug-ins. This should be rare.
+* An `SE_MAJOR` update should be planned in a way that bundles as many pending
+  backwards incompatible changes together as possible to avoid breaking plug-ins
+  multiple times.
+* There will be an announcement giving a grace period before the update happens.
+
+### SE_MINOR
+* Backwards compatible changes.
+  * Adding a new variable, struct, method, enumeration, etc.
+  * Trivial deprecation of a variable, etc. by setting it to a no-op values,
+    e.g., 0 or `NULL`.
+
+### SE_PATCH
+* Backwards compatible bug fixes.
+
+## Conventions
+* Once a member is added to a struct, it cannot be removed, reordered, renamed,
+  or repurposed (i.e., assigned a different functionality).
+* "Renaming" a member is equivalent to adding a new member with a new name and
+  eventually deprecating the member with the old name.
+* Fields that cannot be 0 or `NULL` can be deprecated in a backwards compatible
+  manner by zero-initialization. 
+  * Plug-ins must perform input validation on these fields for 0 and `NULL`
+    before consuming them.
```

Comment:
Should we have another bullet point for fields set by plugin? This case would be more common since most structs we have are populated by plug-in and not TensorFlow. In this case it should be something like:

TensorFlow will validate if any field which should be is deprecated by setting it to `0` or `NULL` is set to a different value and print a warning.

Reply:
You're right. How does "Both core TensorFlow and plug-ins" sounds? Both sides should validate.

Reply:
I feel like its a bit different since TF is responsible for deprecation in both cases.

* If field is set by TF, then plug-in needs to check if it is 0 before using.
* If field is set by plug-in and TF deprecates it, then TF would never use it anyways. So "validate for 0 before consuming" is a noop. Instead, makes more sense if TF checks it is "non-zero" and print warning.

Reply:
True. Added what you said. :)